From cf32692d9a8f9cbef9b27457ed46cb774a356646 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Tue, 21 Nov 2023 11:23:10 +0100
Subject: [PATCH 452/954] feat(stm32mp2): map DDR part non-secure when M33 is
 TDCID

If Cortex-M33 is TDCID, it will configure the DDR firewall before TF-A
has started. And the part where U-Boot and its DT are loaded are set as
non-secure by the firewall. The MMU configuration should then reflect
that. Unmap DDR secure after OP-TEE is loaded, and map non-secure DDR
as non-secure read-write memory.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 6afbb5d934dc429160919b1c55b7b70ca142f211 feat(stm32mp2): map DDR part non-secure when M33 is TDCID]
Change-Id: I3ded707fe76098ab0e2dcbba272bef052995f98a
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/347832
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 1d50f22990..aa7e53f13f 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -593,6 +593,21 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 		}
 		break;
 
+#if STM32MP_M33_TDCID
+	case BL32_EXTRA1_IMAGE_ID:
+		dsbsy();
+		isb();
+		stm32mp_unmap_ddr();
+		err = mmap_add_dynamic_region(STM32MP_BL33_BASE, STM32MP_BL33_BASE,
+					      STM32MP_DDR_MAX_SIZE - (STM32MP_BL33_BASE -
+								      STM32MP_DDR_BASE),
+					      MT_MEMORY | MT_RW | MT_NS);
+		if (err != 0) {
+			panic();
+		}
+		break;
+#endif
+
 	case BL33_IMAGE_ID:
 #if PSA_FWU_SUPPORT
 		stm32_fwu_set_boot_idx();
-- 
2.53.0

